nmt.go

```diff
@@ -418,8 +412,12 @@ func (n *NamespacedMerkleTree) Push(namespacedData namespace.PrefixedData) error
 		return err
 	}
 
+	// compute the leaf hash
+	res := n.treeHasher.MustHashLeaf(namespacedData)
```

Comment:
why not bubble the error here if there ever is one?

Reply:
There are two reasons why the error is not being bubbled up:

- In line 410, the data validity is checked through the `validateAndExtractNamespace` function call. As a result, the `HashLeaf` function call cannot produce an error. To handle this scenario, the `MustHashLeaf` function is introduced, which assumes that the data passed to it has already been validated. This approach is similar to the current situation and avoids any unnecessary error handling.

- If we were to use `HashLeaf` and bubble up the error, we would need to introduce an if block in our code i.e., the `Push` method. However, this if block would not be covered in our tests, leading to a reduction in test coverage percentage. Therefore, I have opted not to bubble up the error in this situation.

Please let me know if this addresses your question.

Reply:
that makes sense, but documenting it in the code for posterity would be a good addition imo. we could also just ignore the error if its impossible (and we have comments), but in general I think its always best to bubble even if we don't strictly need to. Its one less thing the reader or future contributors have to think about.

I'm not blocking on it though! will defer to the author :slightly_smiling_face:

Reply:
Fair point, replaced MustHashLeaf with HashLeaf and bubbled up the error, please see https://github.com/celestiaorg/nmt/pull/172/commits/d60367b11a2a123c4fd06b9772cc30f78793b666

Reply:
Yeah, 100% merge imo